core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -351,11 +366,14 @@ V put(K key, V value, Function<ByteBuffer, V> transformer) {
         if (!c.publish()) {
             c.freeHandle(hi);
             rebalance(c);
-            put(key, value, transformer);
-            return null;
+            return put(key, value, transformer);
         }
 
-        finishAfterPublishing(opData, c);
+        Handle oldHandle = finishAfterPublishing(opData, c);
+
+        if (oldHandle != null) {
+            return transformer != null ? oldHandle.transform(transformer) : null;
```

Comment:
There is still an atomicity problem here, since the value in oldHandle can be deleted meanwhile.

Reply:
You're right, will fix